services-marionette/firefox-services-tests/apps/base.py

```diff
@@ -0,0 +1,88 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+from marionette_driver import expected, Wait
+from marionette_driver.errors import NoSuchElementException
+
+import re
+
+
+class Base(object):
+
+    def __init__(self, marionette):
+        self.marionette = marionette
+        self.CHROME = 'chrome'
+        self.CONTENT = 'content'
+        self.set_context(self.CONTENT)
+
+    def launch(self, url):
+        if url is not None:
+            regex = re.compile(
+                r'^(?:http|ftp)s?://'
+                r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)'
+                r'+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'
+                r'localhost|'
+                r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'
+                r'(?::\d+)?'
+                r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+            if regex.match(url):
+                self.marionette.navigate(url)
+            else:
+                raise ValueError('Url is malformed.')
+        else:
+            raise ValueError("Url must contain a value.")
+
+    def is_element_present(self, by, locator):
+        try:
+            self.marionette.find_element(by, locator)
+            return True
+        except NoSuchElementException:
+            return False
+
+    def is_element_displayed(self, by, locator):
+        try:
+            return self.marionette.find_element(by, locator).is_displayed()
+        except NoSuchElementException:
+            return False
+
+    def wait_for_element_displayed(self, by, locator):
+        return Wait(self.marionette).until(expected.element_displayed(
+            Wait(self.marionette).until(expected.element_present(by, locator))))
+
+    def wait_for_element_present(self, by, locator):
+        Wait(self.marionette).until(expected.element_present(by, locator))
+
+    def wait_for_element_enabled(self, by, locator):
+        Wait(self.marionette).until(
+            expected.element_enabled(lambda m: m.find_element(by, locator)))
+
+    def wait_for_element_not_displayed(self, by, locator):
+        Wait(self.marionette).until(expected.element_not_displayed(
+            Wait(self.marionette).until(expected.element_present(by, locator))))
+
+    def wait_for_element_not_present(self, by, locator):
+        Wait(self.marionette).until(expected.element_not_present(by, locator))
+
+    def wait_for_element_not_enabled(self, by, locator):
+        Wait(self.marionette).until(
```

Comment:
David Burns and I had a looooong conversation about the timeouts. It basically boils down to implicit vs explicit waits. Explicit waits are a better practice. If the wait is not working then there is probably a problem with the way we are querying for the element.